diff --git a/fs/Kconfig b/fs/Kconfig
index 2fb08d05d..70c64078d 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -405,4 +405,11 @@ source "fs/unicode/Kconfig"
 config IO_WQ
 	bool
 
+config NOMOUNT
+	bool "NoMount Path Redirection Subsystem"
+	default y
+	help
+	  NoMount allows path redirection and virtual file injection
+	  without mounting filesystems. Useful for systemless modifications.
+
 endmenu
diff --git a/fs/Makefile b/fs/Makefile
index c6963a9a4..28f0b1854 100644
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -22,6 +22,8 @@ obj-y :=	open.o read_write.o file_table.o super.o \
 		fs_types.o fs_context.o fs_parser.o fsopen.o init.o \
 		kernel_read_file.o mnt_idmapping.o remap_range.o
 
+obj-$(CONFIG_NOMOUNT) += nomount.o
+
 obj-$(CONFIG_BUFFER_HEAD)	+= buffer.o mpage.o
 obj-$(CONFIG_PROC_FS)		+= proc_namespace.o
 obj-$(CONFIG_LEGACY_DIRECT_IO)	+= direct-io.o
diff --git a/fs/d_path.c b/fs/d_path.c
index 5f4da5c8d..e16ad11a3 100644
--- a/fs/d_path.c
+++ b/fs/d_path.c
@@ -246,6 +246,10 @@ static void get_fs_root_rcu(struct fs_struct *fs, struct path *root)
 	} while (read_seqcount_retry(&fs->seq, seq));
 }
 
+#ifdef CONFIG_NOMOUNT
+extern char *nomount_handle_dpath(const struct path *path, char *buf, int buflen);
+#endif
+
 /**
  * d_path - return the path of a dentry
  * @path: path to report
@@ -267,6 +271,13 @@ char *d_path(const struct path *path, char *buf, int buflen)
 	DECLARE_BUFFER(b, buf, buflen);
 	struct path root;
 
+#ifdef CONFIG_NOMOUNT
+	char *nm_path = nomount_handle_dpath(path, buf, buflen);
+	if (unlikely(nm_path)) {
+		return nm_path;
+	}
+#endif
+
 	/*
 	 * We have various synthetic filesystems that never get mounted.  On
 	 * these filesystems dentries are never used for lookup purposes, and
diff --git a/fs/namei.c b/fs/namei.c
index b31b583f1..ed75b39ad 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -126,6 +126,11 @@
 
 #define EMBEDDED_NAME_MAX	(PATH_MAX - offsetof(struct filename, iname))
 
+#ifdef CONFIG_NOMOUNT
+extern struct filename *nomount_handle_getname(struct filename *name);
+extern int nomount_handle_permission(struct inode *inode, int mask);
+#endif
+
 struct filename *
 getname_flags(const char __user *filename, int flags, int *empty)
 {
@@ -201,6 +206,11 @@ getname_flags(const char __user *filename, int flags, int *empty)
 
 	result->uptr = filename;
 	result->aname = NULL;
+#ifdef CONFIG_NOMOUNT
+	if (!IS_ERR(result)) {
+		result = nomount_handle_getname(result);
+	}
+#endif
 	audit_getname(result);
 	return result;
 }
@@ -249,6 +259,11 @@ getname_kernel(const char * filename)
 	memcpy((char *)result->name, filename, len);
 	result->uptr = NULL;
 	result->aname = NULL;
+#ifdef CONFIG_NOMOUNT
+	if (!IS_ERR(result)) {
+		result = nomount_handle_getname(result);
+	}
+#endif
 	atomic_set(&result->refcnt, 1);
 	audit_getname(result);
 
@@ -402,6 +417,12 @@ int generic_permission(struct mnt_idmap *idmap, struct inode *inode,
 {
 	int ret;
 
+#ifdef CONFIG_NOMOUNT
+	int nm_perm = nomount_handle_permission(inode, mask);
+	if (unlikely(nm_perm < 0)) return nm_perm;
+	if (unlikely(nm_perm > 0)) return 0;
+#endif
+
 	/*
 	 * Do the basic permission checks.
 	 */
@@ -506,6 +527,12 @@ int inode_permission(struct mnt_idmap *idmap,
 {
 	int retval;
 
+#ifdef CONFIG_NOMOUNT
+	int nm_perm = nomount_handle_permission(inode, mask);
+	if (unlikely(nm_perm < 0)) return nm_perm;
+	if (unlikely(nm_perm > 0)) return 0;
+#endif
+
 	retval = sb_permission(inode->i_sb, inode, mask);
 	if (retval)
 		return retval;
diff --git a/fs/readdir.c b/fs/readdir.c
index c8c46e294..53c690036 100644
--- a/fs/readdir.c
+++ b/fs/readdir.c
@@ -83,6 +83,9 @@ EXPORT_SYMBOL(wrap_directory_iterator);
 	unsafe_copy_to_user(dst, src, len, label);		\
 } while (0)
 
+#ifdef CONFIG_NOMOUNT
+extern int nomount_handle_iterate_dir(struct file *file, struct dir_context *ctx);
+#endif
 
 int iterate_dir(struct file *file, struct dir_context *ctx)
 {
@@ -103,7 +106,11 @@ int iterate_dir(struct file *file, struct dir_context *ctx)
 	res = -ENOENT;
 	if (!IS_DEADDIR(inode)) {
 		ctx->pos = file->f_pos;
+#ifdef CONFIG_NOMOUNT
+		res = nomount_handle_iterate_dir(file, ctx);
+#else
 		res = file->f_op->iterate_shared(file, ctx);
+#endif
 		file->f_pos = ctx->pos;
 		fsnotify_access(file);
 		file_accessed(file);
diff --git a/fs/stat.c b/fs/stat.c
index 517fa4ed8..9078b57a3 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -141,6 +141,10 @@ int vfs_getattr_nosec(const struct path *path, struct kstat *stat,
 }
 EXPORT_SYMBOL(vfs_getattr_nosec);
 
+#ifdef CONFIG_NOMOUNT
+extern int nomount_handle_getattr(int ret, const struct path *path, struct kstat *stat);
+#endif
+
 /*
  * vfs_getattr - Get the enhanced basic attributes of a file
  * @path: The file of interest
@@ -173,7 +177,11 @@ int vfs_getattr(const struct path *path, struct kstat *stat,
 	retval = security_inode_getattr(path);
 	if (retval)
 		return retval;
+#ifdef CONFIG_NOMOUNT
+    return nomount_handle_getattr(vfs_getattr_nosec(path, stat, request_mask, query_flags), path, stat);
+#else
 	return vfs_getattr_nosec(path, stat, request_mask, query_flags);
+#endif
 }
 EXPORT_SYMBOL_NS(vfs_getattr, ANDROID_GKI_VFS_EXPORT_ONLY);
 
diff --git a/fs/statfs.c b/fs/statfs.c
index 2a37f9f3d..ba9f23b73 100644
--- a/fs/statfs.c
+++ b/fs/statfs.c
@@ -11,6 +11,10 @@
 #include <linux/compat.h>
 #include "internal.h"
 
+#ifdef CONFIG_NOMOUNT
+extern void nomount_spoof_statfs(const struct path *path, struct kstatfs *buf);
+#endif
+
 static int flags_by_mnt(int mnt_flags)
 {
 	int flags = 0;
@@ -90,6 +94,9 @@ int vfs_statfs(const struct path *path, struct kstatfs *buf)
 	error = statfs_by_dentry(path->dentry, buf);
 	if (!error)
 		buf->f_flags = calculate_f_flags(path->mnt);
+#ifdef CONFIG_NOMOUNT
+	nomount_spoof_statfs(path, buf);
+#endif
 	return error;
 }
 EXPORT_SYMBOL_NS(vfs_statfs, ANDROID_GKI_VFS_EXPORT_ONLY);
